File: typescript/src/transaction.ts
import { parseAmountMinor } from './amount.js';
import { domainSeparatedDigest } from './canonical.js';

export interface PostingManifestDigestProfile {
  readonly domain: string;
  readonly contractVersion: string;
}

export const DEFAULT_POSTING_MANIFEST_DIGEST_PROFILE: PostingManifestDigestProfile = Object.freeze({
  domain: 'value-core/posting-manifest',
  contractVersion: 'v1',
});

export interface CanonicalPostingInput {
  readonly accountId: string;
  readonly asset: string;
  readonly amountMinor: string;
}

export interface PostingManifestInput {
  readonly declaredCount: number;
  readonly postings: readonly CanonicalPostingInput[];
  readonly closed: boolean;
  readonly digest?: string;
}

function assertPostingShape(posting: CanonicalPostingInput): bigint {
  if (posting.accountId.length === 0) throw new Error('Posting account is required');
  if (posting.asset.length === 0) throw new Error('Posting asset is required');
  const amount = parseAmountMinor(posting.amountMinor);
  if (amount === 0n) throw new Error('Canonical postings must be non-zero');
  return amount;
}

function totalsByAsset(postings: readonly CanonicalPostingInput[]): ReadonlyMap<string, bigint> {
  const totals = new Map<string, bigint>();
  for (const posting of postings) {
    const amount = assertPostingShape(posting);
    totals.set(posting.asset, (totals.get(posting.asset) ?? 0n) + amount);
  }
  return totals;
}

export function createPostingManifestDigest(
  postings: readonly CanonicalPostingInput[],
  profile: PostingManifestDigestProfile = DEFAULT_POSTING_MANIFEST_DIGEST_PROFILE,
): Promise<string> {
  for (const posting of postings) assertPostingShape(posting);
  return domainSeparatedDigest(
    profile.domain,
    profile.contractVersion,
    postings.map((posting, index) => ({
      postingSequence: index + 1,
      accountId: posting.accountId,
      asset: posting.asset,
      amountMinor: posting.amountMinor,
    })),
  );
}

export function validateBalancedTransaction(postings: readonly CanonicalPostingInput[]): {
  readonly totalMinor: '0';
} {
  if (postings.length < 2) throw new Error('A balanced transaction requires at least two postings');
  for (const [asset, total] of totalsByAsset(postings)) {
    if (total !== 0n) throw new Error(`Posting balance for asset ${asset} is non-zero`);
  }
  return { totalMinor: '0' };
}

export function validateSingleAssetTransaction(input: {
  readonly asset: string;
  readonly postings: readonly CanonicalPostingInput[];
  readonly debitMeansIncrease?: boolean;
}): {
  readonly asset: string;
  readonly postings: readonly CanonicalPostingInput[];
} {
  if (input.debitMeansIncrease === true) {
    throw new Error('Posting sign convention is always the named account perspective');
  }
  if (input.asset.length === 0) throw new Error('Transaction asset is required');
  if (input.postings.some((posting) => posting.asset !== input.asset)) {
    throw new Error('A canonical transaction may name exactly one asset');
  }
  validateBalancedTransaction(input.postings);
  return { asset: input.asset, postings: input.postings };
}

export async function validatePostingManifest(
  input: PostingManifestInput,
  profile: PostingManifestDigestProfile = DEFAULT_POSTING_MANIFEST_DIGEST_PROFILE,
): Promise<PostingManifestInput> {
  if (!Number.isSafeInteger(input.declaredCount) || input.declaredCount < 2) {
    throw new Error('Posting manifest must declare at least two postings');
  }
  if (!input.closed) throw new Error('Posting manifest must be closed before commit');
  if (input.postings.length !== input.declaredCount) {
    throw new Error('Posting manifest count does not match its closed posting set');
  }
  validateBalancedTransaction(input.postings);
  if (input.digest === undefined || !/^[0-9a-f]{64}$/u.test(input.digest)) {
    throw new Error('Posting manifest requires a lowercase SHA-256 digest');
  }
  if (input.digest !== (await createPostingManifestDigest(input.postings, profile))) {
    throw new Error('Posting manifest digest mismatch');
  }
  return input;
}

export interface AccountBalance {
  readonly accountId: string;
  readonly asset: string;
  readonly balanceMinor: string;
  readonly allowNegative?: boolean;
}

/**
 * Applies all posting deltas to a detached balance set. Validation completes before a result is
 * returned, so callers can persist the result atomically without observing partial mutation.
 */
export function applyBalancedTransaction(input: {
  readonly balances: readonly AccountBalance[];
  readonly postings: readonly CanonicalPostingInput[];
}): readonly AccountBalance[] {
  validateBalancedTransaction(input.postings);
  const balances = new Map<string, AccountBalance>();
  for (const balance of input.balances) {
    if (balances.has(balance.accountId)) throw new Error('Account balance identity is duplicated');
    if (balance.accountId.trim().length === 0 || balance.asset.trim().length === 0) {
      throw new Error('Account balance identity and asset are required');
    }
    parseAmountMinor(balance.balanceMinor);
    balances.set(balance.accountId, balance);
  }

  const deltas = new Map<string, bigint>();
  for (const posting of input.postings) {
    const balance = balances.get(posting.accountId);
    if (balance === undefined)
      throw new Error(`Posting account ${posting.accountId} is unavailable`);
    if (balance.asset !== posting.asset)
      throw new Error('Posting asset does not match its account');
    deltas.set(
      posting.accountId,
      (deltas.get(posting.accountId) ?? 0n) + parseAmountMinor(posting.amountMinor),
    );
  }

  return Object.freeze(
    input.balances.map((balance) => {
      const next = parseAmountMinor(balance.balanceMinor) + (deltas.get(balance.accountId) ?? 0n);
      const checked = parseAmountMinor(next.toString());
      if (checked < 0n && balance.allowNegative !== true) {
        throw new Error(`Account ${balance.accountId} has insufficient value`);
      }
      return Object.freeze({ ...balance, balanceMinor: checked.toString() });
    }),
  );
}

/** Creates the literal inverse of a balanced transaction without changing its asset or accounts. */
export function createTransactionReversal(
  postings: readonly CanonicalPostingInput[],
): readonly CanonicalPostingInput[] {
  validateBalancedTransaction(postings);
  const reversal = Object.freeze(
    postings.map((posting) =>
      Object.freeze({
        ...posting,
        amountMinor: (-parseAmountMinor(posting.amountMinor)).toString(),
      }),
    ),
  );
  validateBalancedTransaction(reversal);
  return reversal;
}

File: rust/src/transaction.rs
use serde::Serialize;
use serde_json::to_value;

use crate::amount::parse_amount_minor;
use crate::canonical::domain_separated_digest;
use crate::{ValueError, ValueResult, is_lower_sha256, nonempty};

pub const DEFAULT_POSTING_MANIFEST_DOMAIN: &str = "value-core/posting-manifest";
pub const DEFAULT_POSTING_MANIFEST_CONTRACT_VERSION: &str = "v1";

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct PostingManifestDigestProfile<'a> {
    pub domain: &'a str,
    pub contract_version: &'a str,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CanonicalPosting {
    pub account_id: String,
    pub asset: String,
    pub amount_minor: String,
}

fn validate_posting(posting: &CanonicalPosting) -> ValueResult<()> {
    nonempty(&posting.account_id, "Posting account is required")?;
    nonempty(&posting.asset, "Posting asset is required")?;
    if parse_amount_minor(&posting.amount_minor)? == 0 {
        return Err(ValueError::new("Canonical postings must be non-zero"));
    }
    Ok(())
}
/// # Errors
/// Returns `ValueError` when input violates an authority invariant.
pub fn create_posting_manifest_digest(postings: &[CanonicalPosting]) -> ValueResult<String> {
    create_posting_manifest_digest_with_profile(
        postings,
        PostingManifestDigestProfile {
            domain: DEFAULT_POSTING_MANIFEST_DOMAIN,
            contract_version: DEFAULT_POSTING_MANIFEST_CONTRACT_VERSION,
        },
    )
}

/// # Errors
/// Returns [`ValueError`] when postings or the digest profile are invalid.
pub fn create_posting_manifest_digest_with_profile(
    postings: &[CanonicalPosting],
    profile: PostingManifestDigestProfile<'_>,
) -> ValueResult<String> {
    let mut manifest = Vec::with_capacity(postings.len());
    for (index, posting) in postings.iter().enumerate() {
        validate_posting(posting)?;
        manifest.push(serde_json::json!({
            "postingSequence": index + 1,
            "accountId": posting.account_id,
            "asset": posting.asset,
            "amountMinor": posting.amount_minor,
        }));
    }
    domain_separated_digest(
        profile.domain,
        profile.contract_version,
        &to_value(manifest).map_err(|error| ValueError::new(error.to_string()))?,
    )
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AccountBalance {
    pub account_id: String,
    pub asset: String,
    pub balance_minor: String,
    pub allow_negative: bool,
}

/// Applies a balanced posting set to detached balances without partial mutation.
///
/// # Errors
/// Returns [`ValueError`] when the transaction or any resulting balance is invalid.
pub fn apply_balanced_transaction(
    balances: &[AccountBalance],
    postings: &[CanonicalPosting],
) -> ValueResult<Vec<AccountBalance>> {
    validate_balanced_transaction(postings)?;
    let mut indexed = std::collections::BTreeMap::new();
    for balance in balances {
        nonempty(
            &balance.account_id,
            "Account balance identity and asset are required",
        )?;
        nonempty(
            &balance.asset,
            "Account balance identity and asset are required",
        )?;
        let opening_balance = parse_amount_minor(&balance.balance_minor)?;
        if opening_balance < 0 && !balance.allow_negative {
            return Err(ValueError::new(format!(
                "Account {} cannot start with negative value",
                balance.account_id
            )));
        }
        if indexed
            .insert(balance.account_id.as_str(), balance)
            .is_some()
        {
            return Err(ValueError::new("Account balance identity is duplicated"));
        }
    }
    let mut deltas = std::collections::BTreeMap::<&str, i128>::new();
    for posting in postings {
        let balance = indexed.get(posting.account_id.as_str()).ok_or_else(|| {
            ValueError::new(format!(
                "Posting account {} is unavailable",
                posting.account_id
            ))
        })?;
        if balance.asset != posting.asset {
            return Err(ValueError::new("Posting asset does not match its account"));
        }
        let delta = i128::from(parse_amount_minor(&posting.amount_minor)?);
        let current = deltas.entry(&posting.account_id).or_default();
        *current = current
            .checked_add(delta)
            .ok_or_else(|| ValueError::new("Value arithmetic overflow"))?;
    }
    balances
        .iter()
        .map(|balance| {
            let next = i128::from(parse_amount_minor(&balance.balance_minor)?)
                + deltas
                    .get(balance.account_id.as_str())
                    .copied()
                    .unwrap_or(0);
            let next =
                i64::try_from(next).map_err(|_| ValueError::new("Value arithmetic overflow"))?;
            if next < 0 && !balance.allow_negative {
                return Err(ValueError::new(format!(
                    "Account {} has insufficient value",
                    balance.account_id
                )));
            }
            Ok(AccountBalance {
                account_id: balance.account_id.clone(),
                asset: balance.asset.clone(),
                balance_minor: next.to_string(),
                allow_negative: balance.allow_negative,
            })
        })
        .collect()
}

/// Creates a literal inverse of a balanced transaction.
///
/// # Errors
/// Returns [`ValueError`] when the original transaction is invalid.
pub fn create_transaction_reversal(
    postings: &[CanonicalPosting],
) -> ValueResult<Vec<CanonicalPosting>> {
    validate_balanced_transaction(postings)?;
    let reversal = postings
        .iter()
        .map(|posting| {
            Ok(CanonicalPosting {
                account_id: posting.account_id.clone(),
                asset: posting.asset.clone(),
                amount_minor: parse_amount_minor(&posting.amount_minor)?
                    .checked_neg()
                    .ok_or_else(|| ValueError::new("Value arithmetic overflow"))?
                    .to_string(),
            })
        })
        .collect::<ValueResult<Vec<_>>>()?;
    validate_balanced_transaction(&reversal)?;
    Ok(reversal)
}
/// # Errors
/// Returns `ValueError` when input violates an authority invariant.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct BalancedTransaction {
    pub total_minor: &'static str,
}

/// # Errors
/// Returns [`ValueError`] when postings are empty, malformed, or unbalanced.
pub fn validate_balanced_transaction(
    postings: &[CanonicalPosting],
) -> ValueResult<BalancedTransaction> {
    if postings.len() < 2 {
        return Err(ValueError::new(
            "A balanced transaction requires at least two postings",
        ));
    }
    let mut totals = std::collections::BTreeMap::<&str, i128>::new();
    for posting in postings {
        validate_posting(posting)?;
        *totals.entry(&posting.asset).or_default() +=
            i128::from(parse_amount_minor(&posting.amount_minor)?);
    }
    if let Some((asset, _)) = totals.iter().find(|(_, total)| **total != 0) {
        return Err(ValueError::new(format!(
            "Posting balance for asset {asset} is non-zero"
        )));
    }
    Ok(BalancedTransaction { total_minor: "0" })
}
/// # Errors
/// Returns `ValueError` when input violates an authority invariant.
pub fn validate_single_asset_transaction(
    asset: &str,
    postings: &[CanonicalPosting],
    debit_means_increase: bool,
) -> ValueResult<SingleAssetTransaction> {
    if debit_means_increase {
        return Err(ValueError::new(
            "Posting sign convention is always the named account perspective",
        ));
    }
    nonempty(asset, "Transaction asset is required")?;
    if postings.iter().any(|posting| posting.asset != asset) {
        return Err(ValueError::new(
            "A canonical transaction may name exactly one asset",
        ));
    }
    validate_balanced_transaction(postings)?;
    Ok(SingleAssetTransaction {
        asset: asset.to_owned(),
        postings: postings.to_vec(),
    })
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SingleAssetTransaction {
    pub asset: String,
    pub postings: Vec<CanonicalPosting>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PostingManifest {
    pub declared_count: usize,
    pub postings: Vec<CanonicalPosting>,
    pub closed: bool,
    pub digest: String,
}
/// # Errors
/// Returns `ValueError` when input violates an authority invariant.
pub fn validate_posting_manifest(
    declared_count: usize,
    postings: &[CanonicalPosting],
    closed: bool,
    digest: &str,
) -> ValueResult<PostingManifest> {
    validate_posting_manifest_with_profile(
        declared_count,
        postings,
        closed,
        digest,
        PostingManifestDigestProfile {
            domain: DEFAULT_POSTING_MANIFEST_DOMAIN,
            contract_version: DEFAULT_POSTING_MANIFEST_CONTRACT_VERSION,
        },
    )
}

/// # Errors
/// Returns [`ValueError`] when a posting manifest violates balance, closure, or digest invariants.
pub fn validate_posting_manifest_with_profile(
    declared_count: usize,
    postings: &[CanonicalPosting],
    closed: bool,
    digest: &str,
    profile: PostingManifestDigestProfile<'_>,
) -> ValueResult<PostingManifest> {
    if !(2..=9_007_199_254_740_991).contains(&declared_count) {
        return Err(ValueError::new(
            "Posting manifest must declare at least two postings",
        ));
    }
    if !closed {
        return Err(ValueError::new(
            "Posting manifest must be closed before commit",
        ));
    }
    if postings.len() != declared_count {
        return Err(ValueError::new(
            "Posting manifest count does not match its closed posting set",
        ));
    }
    validate_balanced_transaction(postings)?;
    if !is_lower_sha256(digest) {
        return Err(ValueError::new(
            "Posting manifest requires a lowercase SHA-256 digest",
        ));
    }
    if digest != create_posting_manifest_digest_with_profile(postings, profile)? {
        return Err(ValueError::new("Posting manifest digest mismatch"));
    }
    Ok(PostingManifest {
        declared_count,
        postings: postings.to_vec(),
        closed,
        digest: digest.to_owned(),
    })
}
